File: Net/Patterns/State/After/SoldState.cs
namespace Patterns.State.After
{
    using System;

    internal class SoldState : IState
    {
        private readonly GumballMachine gumballMachine;

        public SoldState(GumballMachine gumballMachine)
        {
            this.gumballMachine = gumballMachine;
        }

        public void Insert()
        {
            Console.WriteLine(MachineMessages.InsertWhenSold);
        }

        public void Eject()
        {
            Console.WriteLine(MachineMessages.EjectWhenSold);
        }

        public void Turn()
        {
            Console.WriteLine(MachineMessages.TurnTwice);
        }

        public void Dispense()
        {
            Console.WriteLine(MachineMessages.DispenseSuccessfully);
            gumballMachine.Count = gumballMachine.Count - 1;
            if (gumballMachine.Count == 0)
            {
                Console.WriteLine(MachineMessages.DispenseTheLastGumball);
                gumballMachine.State = gumballMachine.SoldOut;
            }
            else
            {
                gumballMachine.State = gumballMachine.NoQuarter;
            }
        }
    }
}

File: Java/State/After/src/main/state/SoldState.java
package state;

public class SoldState implements State {
 
    GumballMachine gumballMachine;
 
    public SoldState(GumballMachine gumballMachine) {
        this.gumballMachine = gumballMachine;
    }
       
	public void insertQuarter() {
		System.out.println(MachineMessages.InsertWhenSold);
	}
 
	public void ejectQuarter() {
		System.out.println(MachineMessages.EjectWhenSold);
	}
 
	public void turnCrank() {
		System.out.println(MachineMessages.TurnTwice);
	}
 
	public void dispense() {
		gumballMachine.releaseBall();
		if (gumballMachine.getCount()== 0) {
			gumballMachine.setState(gumballMachine.getSoldOutState());
			System.out.println(MachineMessages.DispenseTheLastGumball);
		} else {
			gumballMachine.setState(gumballMachine.getNoQuarterState());
		}
	}
}


